File: experimental/random/bernoulli_mask/tests/assets/bernoulli_mask.py
#!/usr/bin/env python3
# -*- coding: UTF-8 -*-
# ----------------------------------------------------------------------------
# Copyright (c) 2026 Huawei Technologies Co., Ltd.
# This program is free software, you can redistribute it and/or modify it under the terms and conditions of
# CANN Open Software License Agreement Version 2.0 (the "License").
# Please refer to the License for details. You may not use this file except in compliance with the License.
# THIS SOFTWARE IS PROVIDED ON AN "AS IS" BASIS, WITHOUT WARRANTIES OF ANY KIND, EITHER EXPRESS OR IMPLIED,
# INCLUDING BUT NOT LIMITED TO NON-INFRINGEMENT, MERCHANTABILITY, OR FITNESS FOR A PARTICULAR PURPOSE.
# See LICENSE in the root of the software repository for the full text of the License.
# ----------------------------------------------------------------------------

"""TTK golden and deterministic packed-mask inputs for BernoulliMask."""

import numpy as np


__spec__ = {"bernoulli_mask": "BernoulliMaskTestSpec"}


_DTYPE_MAP = {
    "float16": np.float16,
    "fp16": np.float16,
    "float32": np.float32,
    "fp32": np.float32,
    "double": np.float64,
    "float64": np.float64,
    "fp64": np.float64,
    "uint8": np.uint8,
    "int8": np.int8,
    "int16": np.int16,
    "int32": np.int32,
    "int64": np.int64,
    "bool": np.bool_,
}


def _output_dtype(kwargs):
    output_dtypes = kwargs.get("output_dtypes", ("float32",))
    dtype_name = str(output_dtypes[0]).lower()
    if "bfloat16" in dtype_name or "bf16" in dtype_name:
        try:
            from ml_dtypes import bfloat16
        except ImportError as exc:
            raise RuntimeError(
                "TTK bfloat16 cases require the optional ml-dtypes package"
            ) from exc
        return bfloat16
    try:
        return _DTYPE_MAP[dtype_name]
    except KeyError as exc:
        raise ValueError(
            f"unsupported BernoulliMask output dtype: {dtype_name}"
        ) from exc


class BernoulliMaskTestSpec:
    """Decode each packed byte from least-significant bit to most-significant bit."""

    @staticmethod
    def customize_inputs(mask, **kwargs):
        del kwargs
        pattern = np.array(
            [0x00, 0x01, 0x02, 0x80, 0xA5, 0x5A, 0x7F, 0xFF],
            dtype=np.uint8,
        )
        values = np.resize(pattern, mask.size).reshape(mask.shape)
        return (values,)

    @staticmethod
    def golden(mask, *, output_shape, **kwargs):
        shape = tuple(int(dim) for dim in output_shape)
        elements = int(np.prod(shape, dtype=np.int64)) if shape else 1
        unpacked = np.unpackbits(
            np.asarray(mask, dtype=np.uint8).reshape(-1),
            bitorder="little",
        )[:elements]
        return [unpacked.reshape(shape).astype(_output_dtype(kwargs), copy=False)]

    tolerance = {
        "float16": {"standard": "binary_equal"},
        "float32": {"standard": "binary_equal"},
        "float64": {"standard": "binary_equal"},
        "bfloat16": {"standard": "binary_equal"},
        "uint8": {"standard": "binary_equal"},
        "int8": {"standard": "binary_equal"},
        "int16": {"standard": "binary_equal"},
        "int32": {"standard": "binary_equal"},
        "int64": {"standard": "binary_equal"},
        "bool": {"standard": "binary_equal"},
    }

File: experimental/random/bernoulli_mask/op_api/bernoulli_mask.cpp
/**
 * Copyright (c) 2026 Huawei Technologies Co., Ltd.
 * This program is free software, you can redistribute it and/or modify it under the terms and conditions of
 * CANN Open Software License Agreement Version 2.0 (the "License").
 * Please refer to the License for details. You may not use this file except in compliance with the License.
 * THIS SOFTWARE IS PROVIDED ON AN "AS IS" BASIS, WITHOUT WARRANTIES OF ANY KIND, EITHER EXPRESS OR IMPLIED,
 * INCLUDING BUT NOT LIMITED TO NON-INFRINGEMENT, MERCHANTABILITY, OR FITNESS FOR A PARTICULAR PURPOSE.
 * See LICENSE in the root of the software repository for the full text of the License.
 */

#include "bernoulli_mask.h"

#include "opdev/make_op_executor.h"
#include "opdev/op_dfx.h"
#include "opdev/op_log.h"
#include "opdev/shape_utils.h"

using namespace op;

namespace l0op {
OP_TYPE_REGISTER(BernoulliMask);

aclTensor* BernoulliMask(const aclTensor* mask, aclTensor* out, bool maskAliasesOut, aclOpExecutor* executor)
{
    if (mask == nullptr || out == nullptr || executor == nullptr) {
        OP_LOGE(ACLNN_ERR_INNER_NULLPTR, "BernoulliMask received a null argument.");
        return nullptr;
    }
    L0_DFX(BernoulliMask, mask, out);

    auto outputShape = op::ToShapeVector(out->GetViewShape());
    auto outputShapeArray = executor->AllocIntArray(outputShape.data(), outputShape.size());
    if (outputShapeArray == nullptr) {
        OP_LOGE(ACLNN_ERR_INNER_NULLPTR, "BernoulliMask failed to allocate output_shape.");
        return nullptr;
    }

    const int64_t maskAliasMode = maskAliasesOut ? 1 : 0;
    auto args = op::GetOpArgContext(OP_INPUT(mask), OP_OUTPUT(out), OP_ATTR(outputShapeArray, maskAliasMode));
    if (args == nullptr) {
        OP_LOGE(ACLNN_ERR_INNER_NULLPTR, "BernoulliMask failed to create its argument context.");
        return nullptr;
    }
    auto ret = CreatAiCoreKernelLauncher("BernoulliMask", BernoulliMaskOpTypeId(), executor, args);
    if (ret != ACLNN_SUCCESS) {
        OP_LOGE(ACLNN_ERR_INNER_NULLPTR, "BernoulliMask ADD_TO_LAUNCHER_LIST_AICORE failed.");
        return nullptr;
    }
    return out;
}

const aclTensor* BernoulliMask(const aclTensor* mask, const aclTensor* like, aclOpExecutor* executor)
{
    if (mask == nullptr || like == nullptr || executor == nullptr) {
        return nullptr;
    }
    auto out = executor->AllocTensor(like->GetViewShape(), like->GetDataType(), op::Format::FORMAT_ND);
    if (out == nullptr) {
        OP_LOGE(ACLNN_ERR_INNER_NULLPTR, "BernoulliMask failed to allocate output.");
        return nullptr;
    }
    return BernoulliMask(mask, out, false, executor);
}
} // namespace l0op
